src/firefly/js/visualize/draw/ShapeDataObj.js

```diff
@@ -1002,33 +1020,256 @@ function drawEllipse(drawObj, ctx, plot, drawParams, onlyAddToPath) {
+        if (x === 0) {
+            ra = y >= 0 ? 90 : 270;
+        } else {
+            ra = toDegrees(Math.atan(y / x));     // -90~90
+            if (x < 0) {                          // -90~270
+                ra += 180;
+            }
+            if (ra < 0) ra += 360;
+        }
+
+        return makeWorldPt(ra, dec);
+    };
+
+    // get norm in cartesian coordinate
+    const getNormalFrom = (v1, v2) => {
+        const nx = v1.y * v2.z - v2.y * v1.z;
+        const ny = v1.z * v2.x - v2.z * v1.x;
+        const nz = v1.x * v2.y - v2.x * v1.y;
+
+        const l = Math.sqrt(nx**2 + ny**2 + nz**2);
+        return {x: nx / l, y: ny / l, z: nz / l};
+    };
+
+    /* get the projection oriented boundary on display as moving along the the great circle from pt1 to pt2 */
+    const pt1_c = WorldPtToxyz(plot.getWorldCoords(pt1));
+    const pt2_c = WorldPtToxyz(plot.getWorldCoords(pt2));
+    const norm = getNormalFrom(pt1_c, pt2_c);
+
+    //angle between pt1 & pt2
+    const deg = toDegrees(Math.acos(pt1_c.x * pt2_c.x + pt1_c.y * pt2_c.y + pt1_c.z * pt2_c.z));
+    //angle from vector v to pt2
+    const radToPt2 = toRadians(Math.abs(deg - angleFromCenter));
+
+
+    // find vector v along the greater circle between pt1 & pt2, starting from pt1 with distance angle 'angleFromPt1'
+    // v is solved by mv = y, where dot(v, pt1) = cos(angleFromPt1), dot(v, pt2) = cos(deg-angleFromPt1), dot(n, v) = 0
+    const m = [[pt1_c.x, pt1_c.y, pt1_c.z],
+        [pt2_c.x, pt2_c.y, pt2_c.z],
+        [norm.x, norm.y, norm.z]];
+
+    const y = [Math.cos(toRadians(angleFromCenter)), Math.cos(radToPt2), 0];  // 1: inner product(ret, pt2), 0: inner product(ret, pt1) = 0 (cos(90))
+
+    const det_m = m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]) +
+        m[0][1] * (m[1][2] * m[2][0] - m[2][2] * m[1][0]) +
+        m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
+
+    const adj_m = [[(m[1][1] * m[2][2] - m[2][1] * m[1][2]), -(m[0][1] * m[2][2] - m[2][1] * m[0][2]), (m[0][1] * m[1][2] - m[1][1] * m[0][2])],
+        [-(m[1][0] * m[2][2] - m[2][0] * m[1][2]), (m[0][0] * m[2][2] - m[2][0] * m[0][2]), -(m[0][0] * m[1][2] - m[1][0] * m[0][2])],
+        [(m[1][0] * m[2][1] - m[2][0] * m[1][1]), -(m[0][0] * m[2][1] - m[2][0] * m[0][1]), (m[0][0] * m[1][1] - m[1][0] * m[0][1])]];
+
+
+    const v = adj_m.map((r) => {
+        return (r[0] * y[0] + r[1] * y[1] + r[2] * y[2]) / det_m;
+    });
+
+    return  xyzToWorldPt(v[0], v[1], v[2]);
+}
+
+function getEdgePtOnGreatCircleFromCenterTo(pt, plot) {
+    const eastAngle = 89.99999;
+
+    return getWorldPtByAngleFromProjectCenter(pt, plot, eastAngle);
+}
+
+/**
+ * draw polygon - draw outline of the polygon or fill the polygon
+ * @param drawObj
+ * @param ctx
+ * @param plot
+ * @param drawParams
+ * @param onlyAddToPath
+ */
+function drawPolygon(drawObj, ctx,  plot, drawParams, onlyAddToPath) {
+    const {style, color, lineWidth=1} = drawParams;
+    const isMocTile = drawObj.mocInfo;
```

Comment:
We have always tried to keep the drawing code separate from the business logic. For example this file knows how to draw shapes.  It is necessary to have knowledge of MOC at this level.  Can it not stay at the drawing layer level?  I would like you to consider this and see.  If you think it must know about MOC at such a low level the we will live with it but it does break the paradigm

Reply:
Good point. The logic can be fixed in generic way.